app/models/manageiq/providers/awx/automation_manager.rb

```diff
@@ -25,7 +25,7 @@ class << self
   belongs_to :provider, :autosave => true, :dependent => :destroy
   before_validation :update_provider_zone
 
-  after_save :change_maintenance_for_provider, :if => proc { |ems| ems.saved_change_to_enabled? }
+  after_save :change_maintenance_for_provider, :if => proc { |ems| ems.saved_change_to_enabled? && !ems.enabled_before_last_save.nil? }
```

Comment:
@agrare thoughts on this?  Do we think attribute defaults are worthwhile doing if they trigger "changed"?  This is tough to track down because you need to find anyplace that is checking for changes on the attribute where it's not expecting the setting to the default value will trigger changes.

Reply:
Hm could we use a sql default instead of a rails one?

Reply:
Yes, good point.  That is another option to try before going back to default_value_for.

Reply:
So, it works if we undo this PR, drop the default attribute and use database defaults:


```diff --git a/app/models/ext_management_system.rb b/app/models/ext_management_system.rb
index 7f96a31626..b3ff4c50fb 100644
--- a/app/models/ext_management_system.rb
+++ b/app/models/ext_management_system.rb
@@ -306,8 +306,6 @@ class ExtManagementSystem < ApplicationRecord
   alias_method :clusters, :ems_clusters # Used by web-services to return clusters as the property name
   alias_attribute :to_s, :name

-  attribute :enabled, :default => true
-
   # Move ems to maintenance zone and backup current one
   # @param orig_zone [Integer] because of zone of child manager can be changed by parent manager's ensure_managers() callback
   #
```

migration:
```ruby
class AddColumnDefaults < ActiveRecord::Migration[6.1]
  def change
    change_column_default :ext_management_systems, :enabled, from: nil, to: true
  end
end
```

The problem is lots of our default attributes are subclasses so we can't use database defaults...

```
app/models/manageiq/providers/cloud_manager/template.rb:  attribute :cloud, :default => true
app/models/manageiq/providers/cloud_manager/vm.rb:  attribute :cloud, :default => true
app/models/manageiq/providers/infra_manager/template.rb:  attribute :cloud, :default => false
app/models/manageiq/providers/infra_manager/vm.rb:  attribute :cloud, :default => false
app/models/manageiq/providers/physical_infra_manager/vm.rb:  attribute :cloud, :default => false
```

Reply:
default_value_for handles intentional nil values, whereas the change_column_default would have to be more sophisticated to deal with that case I think.  I'm not sure how much we _actually_ rely on that behavior though.

Reply:
Yeah I always lean towards SQL default since model defaults is one of those things that preclude pure-sql concurrent-safe-batch saving

Reply:
We stayed away from it for a long time, but I'd be on board with using it - the main thing in the past where we got bitten was if you had a sql default, then did a Model.new, then the model would _not_ have that value in memory, however it's possible that Rails fixed that.  If so, that makes me feel even more confident about moving to SQL defaults.

Reply:
Yeah, I think the only problem with the sql defaults is the ruby side defaults we're doing such as the `cloud` attribute above where the default changed based on the subclass.  Of course, anything that was deciding defaults beyond a hardcoded value would also be changed... such as moving from constants to hardcoded values such as:

```
endpoint.rb:  attribute :verify_ssl, :default => OpenSSL::SSL::VERIFY_PEER
miq_group.rb:  attribute :group_type, :default => USER_GROUP
```

Or blocks:

```
miq_group.rb:  attribute :sequence, :default => -> { next_sequence }
```

Reply:
So attribute default works for anything that doesn't need to derive defaults based on other attributes (no instance yielded to the block).  It gets tripped up by anything checking "changes" and could be tough to diagnose, such as this PR.  I have a feeling we'll still need default_value_for for some things so I'm trying to get access to do a release of it.

sql default seems like a no brainer for hardcoded values but is more restrictive as you can't use constants like above so those values would need to be in the database default so it's more work to change.  It does work better with change detection though.  It also doesn't handle the allow nil or don't allow nil mechanics as default_value_for did so if you set defaults, it would only be for new records unless you also updated existing records to new defaults.

Reply:
I think Rails works now properly...example:

```ruby
# db.schema.rb
  create_table "chargeback_rate_details", force: :cascade do |t|
    t.boolean "enabled", default: true

# rails c
> ChargebackRateDetail.new
=>
#<ChargebackRateDetail:0x000000010d9bc748
 id: nil,
 enabled: true,
 ...
>
```

Reply:
> I think Rails works now properly...example:

Yes, it works with the limitations I mentioned above.  I'm not sure what to do about STI since some classes set different defaults and we're still stuck with default_value_for for the cases where we need instances provided in the block to derive the default.

Reply:
>  I'm not sure what to do about STI since some classes set different defaults

I'm curious how many of those we have, the `:cloud => true` is probably pretty easily removed since we can just do an STI subclass check.

Reply:
> > I'm not sure what to do about STI since some classes set different defaults
> 
> I'm curious how many of those we have, the `:cloud => true` is probably pretty easily removed since we can just do an STI subclass check.

It might just be `cloud`.  Maybe `default_value_for :name, "No Availability Zone"`. I haven't seen if any classes in the ancestry use a different name base on the class.

Reply:
> Maybe default_value_for :name, "No Availability Zone"

That's a record created by inventory refresh so we can set that value in the parser instead

Reply:
agreed: revert and use the database (disclaimer: I'm biased towards database default values)